### DISCUSSION TITLE
How to get multiple I/O port outputs like 0x80, 0x81, 0x82, 0x83

### DISCUSSION THREAD

**Discussion by hanetzer** [MrGreensWorkshop/RasPiPicoSDK-PicoBiosPostCodeReader#4] [How to] [answered]
Hello. So, I'm doing some tinkering with a ryzen motherboard (Asrock X370 Killer Sli A/C with a ryzen 3200g)
in relation to the coreboot project. Maybe it will go somewhere, maybe not.
In any case, these machines have a 32-bit post code system, which are output to ports 0x80-0x83.
How much issue would it be to be able to sample a 'range' of io ports with this hardware setup?

**Comment by hanetzer** [comment]
changing the 0x80 to either of the other ports 'works' but I've no idea how to get all of it working that way,
and would be a great help to my current project.
so an ideal readout would be 0x83:0x82:0x81:0x80's data, in that order.

**Comment by MrGreensWorkshop** [comment]
Hi, As you already know, the implementation allows you to use only one port at a time.

There are several approaches to achieving this. (ish..)
1) I actually explained how the PIO program works [here](https://www.youtube.com/watch?v=C4Kq1r9A0k0&t=602s),
There is a part that gets the address and checks if it matches the input address.

check out the https://github.com/MrGreensWorkshop/RasPiPicoSDK-PicoBiosPostCodeReader/discussions/4#discussioncomment-6726894

2) Since the Raspberry Pi Pico has two PIOs, one more I/O port can be implemented. 2 ports at most. For 4 I/O ports, you can use 2 Raspberry Pi Picos that are connected to the LPC bus. Also, PIO doesn't allow you to use the same pins, so LED pins should be changed. (If you don't use LEDs, just ignore this.)

3) Exporting I/O addresses to GPIOs, checking with logic ICs (like ISA cards), and inputting that result to the PIO. The thing is, we need time between clocks. I'm not sure if there is a time to do this.

I think that's all I can say. Good luck.

**Reply by hanetzer** [reply]
well, its still an issue of collating the data points together. like, given that the nibble loop count bit works, is there a means one can output, say, addr:data easily enough? (and I'm assuming that filter.val.addr should be set to 0x80 to get 0x8x addresses?)

**Reply by MrGreensWorkshop** [reply]
I did update the answer. Sorry, it's my bad. With this, you can get the address + data pair on the console.

**Reply by hanetzer** [reply]
well, removing the isr shift and changing the printf specifier gives me data like `0x07080020`
(so, only port 0x80), but defining NIBBLE_LOOP_COUNT to 3 ends up with nothing coming out at all

**Reply by MrGreensWorkshop** [reply]
I did update the answer again. Sorry.

**Reply by hanetzer** [reply]
Even after adding the clock wait, still get no output at all, other than the "..." I'm alive signal.
But this is a start, thank you.

**Reply by hanetzer** [reply]
Better, with the added `in pins, 4`, I think. Still only shows as `xx080020`, where the non-xx
digits do not change.
current changes:
```diff
diff --git a/src/lpc_bus_sniffer.pio b/src/lpc_bus_sniffer.pio
index b1ac93e..5d661a6 100644
--- a/src/lpc_bus_sniffer.pio
+++ b/src/lpc_bus_sniffer.pio
@@ -1,7 +1,7 @@
 .program lpc_bus_sniffer
 ;.side_set 1 opt                 ; [DEBUG] side_set <count> (opt) (pindirs) 
 
-.define NIBBLE_LOOP_COUNT 4
+.define NIBBLE_LOOP_COUNT 3
 
 pull                             ; Get the filter from TX FIFO to the OSR.
 mov Y, OSR                       ; Copy OSR to Y.
@@ -32,6 +32,9 @@ in pins, 4                       ; Read the LAD[0-3] pins.
 wait 1 pin 4                     ; Wait until the LCLK pin goes high.
 jmp X-- read_nibble1             ; If X is not equal to zero, decrement it and jump.
 
+wait 0 pin 4                     ; Wait until the LCLK pin goes low.
+in null, 4
+wait 1 pin 4                     ; Wait until the LCLK pin goes high.
 ; Check the frame, START + CY/DR + ADDR[0-3]
 mov X, ISR                       ; Copy ISR to X.
 jmp X!=Y entry_point             ; If X and Y registers are not equal jump.
@@ -46,7 +49,7 @@ wait 0 pin 4                     ; Wait until the LCLK pin goes low.
 in pins, 4                       ; Get the DATA[1].
 wait 1 pin 4                     ; Wait until the LCLK pin goes high.
 
-in null, 24                      ; Shift the ISR 24 bit to get DATA.
+;in null, 24                      ; Shift the ISR 24 bit to get DATA.
 mov OSR, ~ISR                    ; Invert ISR and copy to OSR.
 out pins, 8                      ; Output the DATA to the LEDs.
 push                             ; Push to get the DATA value on terminal.
@@ -117,4 +120,4 @@ void lpc_bus_sniffer_program_init(PIO pio, uint sm, uint offset, uint lpc_bus_pi
    pio_sm_init(pio, sm, offset, &c);
 }
 
-%}
\ No newline at end of file
+%}
diff --git a/src/main.c b/src/main.c
index 3dcdb3a..f07bce2 100755
--- a/src/main.c
+++ b/src/main.c
@@ -6,7 +6,7 @@
 #include "lpc_bus_sniffer.pio.h"
 
 // Set the system frequency to 33 MHz x 6 to make time for handling LPC frames.
-#define SYS_FREQ_IN_KHZ (198 * 1000)
+#define SYS_FREQ_IN_KHZ (264 * 1000)
 
 // LAD[0-3] + LCLK + LFRAME, which starts from GPIO0.
 #define LPC_BUS_PIN_BASE (0U)
@@ -153,7 +153,7 @@ int main() {
     while(1) {
         if (!pio_sm_is_rx_fifo_empty(_pio0, sm_lpc_bus_sniffer)) {
             uint32_t rxdata = _pio0->rxf[sm_lpc_bus_sniffer];
-            printf("data: %02x\n", rxdata);
+            printf("data: %08x\n", rxdata);
         }
     }
```

**Reply by MrGreensWorkshop** [reply]
I did some debugging. Filtering 0x8X works, but it still doesn't give you the address as output, which is not acceptable. I managed to get the filter working and getting the last address as output, so I'm putting it in a branch.

**Comment by MrGreensWorkshop** [comment, marked as answer]
I put the working code on this [branch](https://github.com/MrGreensWorkshop/RasPiPicoSDK-PicoBiosPostCodeReader/tree/getting_all_data_from_address_0x8x).

This is the log output

```
STARTED
initializing the lpc bus sniffer program
Filter: 0x80020000
program loaded at 1, sm: 0
System clock is 264000 kHz
Starting to sample the POST codes.
...
...
data: 0x92180020
...
data: 0x12080020
```

on Linux 
```
# Write 0x92 to address 0x81 (129)
~# echo -e "\x92" | dd of=/dev/port bs=1 count=1 seek=129
```

serial output : `data:  92 18 00 20`
92 -> 0x92 data
18-> 0x81 address (swapped nibbles)

```
# Write 0x12 to address 0x80 (128)
~# echo -e "\x12" | dd of=/dev/port bs=1 count=1 seek=128
```

serial output : `data:  12 08 00 20`
12 -> 0x12 data
08-> 0x80 address (swapped nibbles)

**Reply by hanetzer** [reply]
yes this works quite well. Takes a bit of post-processing to make sense of the data (generally the firmware seems to write port 0x80 thru 0x83, in that order, in sequence) but this appears to dump most of it. So it seems my boot failure is happening due to a null pointer, thanks a lot.

**Reply by MrGreensWorkshop** [reply]
I'm glad to help you diagnose the problem. I hope you can fix it. Thanks for the info about byte order, in case I dive into coreboot.

**Reply by hanetzer** [reply]
I can't say this is true for all boards, but it is the observed behavior of this as seen
through a logic analyzer, and possibly all ryzen machines.

**Comment by MaryJosephine123** [comment]
Our testing platform has a 4-digit 7 segment LED display for displaying the post code. The first two digits are from port 81, and the last two digits are from port 80. Using this github (git clone https://github.com/MrGreensWorkshop/RasPiPicoSDK-PicoBiosPostCodeReader.git), we were able to capture the last two digits (port 80). Can you help us capture all four digits, including port 81? 
I tried this git (git clone https://github.com/MrGreensWorkshop/RasPiPicoSDK-PicoBiosPostCodeReader.git) but i am getting below output: (not port80 or port81)
0x00000020
 0x00000020
 0x00000020
 0x00000020
 0x00000020
 0x00000020
 0x00000020
**We need only port81 data followed by port80. can you please help us by providing the code sir.**

**Comment by MaryJosephine123** [comment]
Hi,
Could you please help us to get only port80 and port81 data??
